### ISSUE TITLE
local multicast/MDNS for relay

### ISSUE THREAD

**Issue by warner** [warner/magic-wormhole#48]
A friend and I were able to transfer a file directly, when both of us were normally using a VPN (which means the data has to go through the relay, very slow), by switching to a local ad-hoc network. (on OS-X, this is done by choosing "Create Network.." from the wifi menu). To accomplish this, I had to run a relay locally, and he had to use `--relay-url=ws://MY-IP-ADDR:4000/v1`. Neither of us were connected to the real internet while on the ad-hoc network, so the normal relay wasn't available.

It'd be slick to have a mode that does this without running a local relay. MCT and I were thinking about this a while ago. I'm imagining something like this:

* use `--relay-url=multicast`, or `--relay-multicast`, or MDNS, or some flag to say "use the local network instead of the default relay server"
* replace the rendezvous protocol client with one that joins a multicast group, sends periodic messages to the group until the other client responds, then sends messages directly to the other client

There might be a cheaper/easier/less-coding way to go, like by having a distinguished side (maybe the sender) run a full rendezvous server (with an in-memory database instead of `relay.sqlite` on disk), then register it on a well-known MDNS/zeroconf/bonjour name. The receiver could repeatedly try to connect to a relay at that well-known name, and then speak the normal relay protocol.

**Comment by takluyver** [comment]
For reference, zget is a similar CLI tool to transfer files, which currently works only over the local network and without encryption (though encryption is being worked on). The sender uses zeroconf MDNS to advertise the file, and the recipient waits for the advertisement to tell it where to connect.

Cross-pinging @nils-werner, author of zget. It looks like these two tools are gradually converging.

For the CLI, it would be good to name the option something like `--local` or `--nointernet` that matches what the user wants to do, rather than something based on the technical details of how it works.

**Comment by warner** [comment]
Excellent, thanks for the pointer!

**Comment by warner** [comment]
Talking with @saljam (a Camlistore guy) today, we had a few thoughts:

* if you can't talk to the normal relay, start a local one and advertise it via MDNS
* https://pypi.python.org/pypi/zeroconf looks useful here.
* the sender should poll all local relays, build up a list of channels currently in use, then allocate a new one, and add it to the local server
* the receiver should poll all local relays, do tab-completion over the union of all channels being advertised, then use the relay advertising the channel selected by the wormhole code
* this is easier than trying to advertise a channel directly to MDNS, since we already have the server code in each client, and it's easier to poll for local relays (`_tcp._wormhole.local`) than to poll for an (unbounded) number of channels (`3._wormhole.local`, `4._wormhole.local`, etc).
* client and server could both run relays, publish a random ID, let the higher ID drive. But it's probably safer to have the sender run the relay, since they're allocating the channel. The "offline code" use case might need to be handled specially.

**Comment by takluyver** [comment]
Are race conditions a concern for picking new channel IDs?

**Comment by warner** [comment]
Yeah, although if the rate of client allocations is low enough, the chances of a collision should be too low to worry about.

With just one (central) rendezvous server, allocations are sound (there's no way for two clients to be allocated the same value) regardless of how fast they arrive. If you aren't on the internet, so you're running a rendezvous server yourself, then you might overlap with someone else who's doing the same thing (running `wormhole send` at the same time as you), and it'd be good to avoid collisions. The "pre-flight" check would be to query all the (MDNS) servers we can find, and ask them for all the channels they're currently hosting. Then the race window opens, and we pick a channel at random, and before we reveal it to the user (in a wormhole code), we stall a little bit for other senders to query us. Say we wait one second. If someone else picked the same channel as we did, and queried us just before we picked that channel, then we'll have a collision.

Eyeballing it, the chances of a collision are the round trip time, divied by the mean inter-arrival request time, multiplied by the probability of choosing the same channel (e.g. 10% for one-digit channels). It's been a long time since I studied probability theory, though, so I wouldn't bet on my math :).

**Comment by warner** [comment]
More thoughts:

We've got two possible workflows. The first is where one side allocates the channel and creates the code, then the other side learns the code and types it in. In this workflow, the server picks the channel (in response to a "please allocate me one" message from the server), and the second side can be confident that the channel exists already.

The second workflow is where the two humans choose the channel ID (along with the rest of the code) randomly, offline, then they go home and type the codes into their computers. In this one, we don't know which side will type it in first. So both sides must be prepared to be the first one, which means they'll claim the channel (by telling the server)

`wormhole send` knows whether it's in the online-allocation or offline-allocation case (the latter is `wormhole send --code=CODE`). `wormhole receive` does not. Tab-completion on the channel ID is not always useful when doing offline codes (if the sender is already running, then the channel-id is claimed and will show up when the receiver asks for the current list, but if the receiver went first, it won't), and anyways it's unlikely that the humans used the same wordlist.

* We need to decide whether the local-network setting should handle the offline flow. It's not obviously useful, but it might be cleaner if all settings handle all workflows.

* A local-network offline-code situation means that both sides must run a relay server, then check for other relays, then check for the requested channel being claimed on any of them. If the channel isn't already present, the side (either sender or receiver) should claim it locally, then wait for the other side to connect and query the channel ID.

* In no case should we claim a channel on an external server that isn't the central server, because that server might actually be another client (involved in an unrelated transfer), which might shut down before our own connection is complete.

**Comment by nils-werner** [comment]
Could it be possible to leave out the relay server entirely? Instead, the sender could advertise the channel ID directly using MDNS, and offer a relay-server-like API to do the remaining negotiation.

I understand that this turns a lot of the system on its head and is probably not easily implemented. However this is the remaining feature about zget, that I prefer: You don't need a central server run by someone else on the internet.

**Comment by warner** [comment]
Yeah, I think that'd work, as long as you know ahead of time that you're both on the same broadcast domain (or whatever gets you a shared MDNS namespace). Maybe `wormhole send --mdns-only foo.zip`. You'd still need some kind of channel-id to avoid colliding with anyone *else* who was doing MDNS-only wormhole at the same time, but I think MDNS has plenty of support for claiming a name and checking to see if anyone else is using it first, so the channel-id could almost always be just "1". I love the idea of avoiding a central server when you can.

Does zget have any way to get a file to someone on a non-local network? Is it MDNS-only?

**Comment by nils-werner** [comment]
I would be cool if wormhole defaulted to first trying a local broadcast and only after some timeout go to the relay server. But again, that adds yet another layer of complexity on top.

> MDNS has plenty of support for claiming a name and checking to see if anyone else is using it first.

From what I've learned doing `zget`, is that zeroconf is really weird about certain things. The Python package does indeed raise an exception if the service name is already taken, but I wouldn't rely on it really. Also, if you kill your process before unregistering the service, you might collide with your previous self, too. It is a really messy system and without a central authority I would simply assume everybody else is doing it wrong and design for the worst case :-)

Also, only using channel 1 would probably make it really likely to have collisions at big networked events like [33c3](https://events.ccc.de/) when everybody is using

    1._magicwormhole._tcp.local.

How are channels allocated on the rendezvous server? Just pick the lowest available one? Could you just use a random function to pick any MDNS channel from 0-20?

In any case, it would be trivially easy for somebody to just DoS the broadcast domain and force a fallback to the rendezvous server by squatting all service names.

> Does zget have any way to get a file to someone on a non-local network? Is it MDNS-only?

Yes, it is MDNS only, doesn't have a rendezvous or relay server and only allows transfers if you are on the same LAN. `zget` does peer finding a bit differently; the equivalent of the channel ID is the SHA-1 hash of the filename you are trying to transfer, which is known to both parties. The sender announces

    <sha-1-of-filename>._zget._http._tcp.local.

and the recipient looks for that. That obviously gives zero protection from attackers but at least makes it unlikely for an accidental collision to happen.

**Comment by warner** [comment]
> How are channels allocated on the rendezvous server? Just pick the lowest available one?

Pretty close. The [allocation code](https://github.com/warner/magic-wormhole/blob/0.9.2/src/wormhole/server/rendezvous.py#L179) chooses a random single-digit code, if any are available. If not, it chooses a random two-digit code. If those are all used up, it does the same for three-digit codes. If *those* are used up too, it just starts picking randomly from 1000-1000000 for a thousand trials and hopes one of them is free.

I wanted to make it slightly more random than "always lowest available", so that people didn't get into the habit of assuming that codes always started with `1-`.


> Could you just use a random function to pick any MDNS channel from 0-20?

Yeah, that would probably work. (we treat "0" as special, for an unauthenticated no-code-necessary throw-caution-to-the-wind mode). If we know that we're being asked to allocate a code (e.g. `wormhole send` without the explicit `--code=` argument), then it'd be nice to use MDNS to query for which channels are already in use, to reduce the chance of collisions somewhat (not entirely, because of races). But if the range we're picking from is large enough, and the rate at which people are using them is low enough (on a given broadcast domain), then it probably doesn't matter.

> In any case, it would be trivially easy for somebody to just DoS the broadcast domain and force a fallback to the rendezvous server by squatting all service names.

The same DoS attack is easy to do against the rendezvous server, so no real change there.
